// mariadb2/old.js
const maria = require('mariadb')
const { EventEmitter } = require('events')
class Client extends EventEmitter {
  constructor(conf) {
    super()
    if (conf) {
      this.pool = new maria.createPool({
        user: "root",
        host: "localhost",
        database: "tachi",
        password: "tachi",
        port: 3306,
        connectionLimit: conf
      })
    } else {
      this.pool = new maria.createPool({
        user: "root",
        host: "localhost",
        database: "tachi",
        password: "tachi",
        port: 3306
      })
    }
    this.pool.on('error', (err, client) => {
      console.error('Unexpected error from the database', err)
      process.exit()
    })
    this.query = (data, cb) => {
      this.pool.getConnection().then(conn => {
        conn.query(data).then(dout => {
          cb(null, dout)
          conn.end()
        }).catch(err => {
          conn.end()
          cb(err, null)
        })
      })
    }
  }
}

module.exports = {
  port: 3306,
  password: "tachi",
  user: "root",
  Client
}
